sound/soc/intel/boards/sof_ssp_common.c

```diff
@@ -96,6 +96,164 @@ enum sof_ssp_codec sof_ssp_detect_amp_type(struct device *dev)
 }
 EXPORT_SYMBOL_NS(sof_ssp_detect_amp_type, SND_SOC_INTEL_SOF_SSP_COMMON);
 
+int sof_ssp_card_late_probe(struct snd_soc_card *card)
+{
+	return sof_ssp_hdmi_card_late_probe(card);
+}
+EXPORT_SYMBOL_NS(sof_ssp_card_late_probe, SND_SOC_INTEL_SOF_SSP_COMMON);
+
+static struct snd_soc_dai_link_component dmic_component[] = {
+	{
+		.name = "dmic-codec",
+		.dai_name = "dmic-hifi",
+	}
+};
+
+static struct snd_soc_dai_link_component platform_component[] = {
+	{
+		/* name might be overridden during probe */
+		.name = "0000:00:1f.3"
+	}
+};
+
+int sof_ssp_set_dmic01_link(struct device *dev, struct snd_soc_dai_link *link,
+			    int be_id)
+{
+	struct snd_soc_dai_link_component *cpus;
+
+	dev_dbg(dev, "link %d: dmic01\n", be_id);
+
+	/* link name */
+	link->name = "dmic01";
+
+	/* cpus */
+	cpus = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component),
+			    GFP_KERNEL);
+	if (!cpus)
+		return -ENOMEM;
+
+	cpus->dai_name = "DMIC01 Pin";
+
+	link->cpus = cpus;
+	link->num_cpus = 1;
+
+	/* codecs */
+	link->codecs = dmic_component;
+	link->num_codecs = ARRAY_SIZE(dmic_component);
+
+	/* platforms */
+	link->platforms = platform_component;
+	link->num_platforms = ARRAY_SIZE(platform_component);
+
+	link->id = be_id;
+	link->init = sof_ssp_dmic_init;
+	link->ignore_suspend = 1;
+	link->no_pcm = 1;
+	link->dpcm_capture = 1;
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sof_ssp_set_dmic01_link, SND_SOC_INTEL_SOF_SSP_COMMON);
+
+int sof_ssp_set_dmic16k_link(struct device *dev, struct snd_soc_dai_link *link,
+			     int be_id)
+{
+	struct snd_soc_dai_link_component *cpus;
+
+	dev_dbg(dev, "link %d: dmic16k\n", be_id);
+
+	/* link name */
+	link->name = "dmic16k";
+
+	/* cpus */
+	cpus = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component),
+			    GFP_KERNEL);
+	if (!cpus)
+		return -ENOMEM;
+
+	cpus->dai_name = "DMIC16k Pin";
+
+	link->cpus = cpus;
+	link->num_cpus = 1;
+
+	/* codecs */
+	link->codecs = dmic_component;
+	link->num_codecs = ARRAY_SIZE(dmic_component);
+
+	/* platforms */
+	link->platforms = platform_component;
+	link->num_platforms = ARRAY_SIZE(platform_component);
+
+	link->id = be_id;
+	link->ignore_suspend = 1;
+	link->no_pcm = 1;
+	link->dpcm_capture = 1;
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sof_ssp_set_dmic16k_link, SND_SOC_INTEL_SOF_SSP_COMMON);
```

Comment:
no, this doesn't belong in ssp common.

Reply:
Maybe my naming is confusing. My intention is moving common codes in machine drivers (sof_rt5682, sof_da7219, sof_cs42l42, sof_nau8825, sof_ssp_amp) to a common module. Maybe I should rename the sof_ssp_common.c to something like sof_intel_board_common.c ?